sdks/java/harness/src/main/java/org/apache/beam/fn/harness/logging/BeamFnLoggingMDC.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.logging;
+
+/** Mapped diagnostic context to be consumed and set on LogEntry protos in BeamFnLoggingClient. */
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class BeamFnLoggingMDC {
+  private static final InheritableThreadLocal<String> instructionId =
+      new InheritableThreadLocal<>();
```

Comment:
```suggestion
  private static final ThreadLocal<String> instructionId =
      new ThreadLocal<>();
```

Inheritable only works if the scope of the child thread is to the instruction that is creating it and if a thread was created in one instruction and is used for multiple instructions then the logged instruction id will be all associated with the original instruction and not any future ones interacting with it.

e.g. `thread-1` for `instruction-A` creates `thread-2`, then `thread-1` finishes `instruction-1` and starts processing `instruction-2` which interacts with `thread-2`, log entries from `thread-1` will be associated with `instruction-1`. Similarly if `thread-3` comes along and interacts with `thread-2` then the logs for `thread-2` will be for `instruction-1`. 

I suggest the `ThreadLocal` since it is easier for people to see that there are logs that aren't associated with an instruction id and for them to learn to manage the `BeamFnLoggingMDC` in child threads then it is for them to find that the wrong instruction id is associated with their logs and to manage the `BeamFnLoggingMDC`.

Reply:
Makes sense. Done